File: leetcode/49-group-anagrams/main.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// 1st attempt: hashtable
// 1 pass for iterating the words and put the words into corresponding hashtable
// 1 pass for grouping the values from hashtable into the result
// Time O(n*klogk) n:number of words, k:length of charactors, klogk is due to the sorting
// Space O(k*n)
// beats 96.55%
func groupAnagrams(strs []string) [][]string {
	hash := make(map[string][]string)
	for i := 0; i < len(strs); i++ {
		word := strs[i]
		key := SortString(strs[i])
		if _, x := hash[key]; x {
			hash[key] = append(hash[key], word)
		} else {
			hash[key] = []string{word}
		}
	}
	result := [][]string{}
	for key := range hash {
		result = append(result, hash[key])
	}
	return result
}

func SortString(w string) string {
	s := strings.Split(w, "")
	sort.Strings(s)
	return strings.Join(s, "")
}

// 2nd attempt: hashtable but base on charactor count
// although it doesnt sort(nlogn), it runs slower than the 1st due to the string manipulation
// Time O(n*2m) n:number of words, k:length of charactors, klogk is due to the sorting
// Space O(k*n)
// beats 89.66%
func groupAnagrams1(strs []string) [][]string {
	hash := make(map[string][]string)
	for i := 0; i < len(strs); i++ {
		word := strs[i]
		key := charCount(strs[i])
		if _, x := hash[key]; x {
			hash[key] = append(hash[key], word)
		} else {
			hash[key] = []string{word}
		}
	}
	result := [][]string{}
	for key := range hash {
		result = append(result, hash[key])
	}
	return result
}

func charCount(s string) string {
	charCnt := [26]int{}
	for i := 0; i < len(s); i++ {
		charCnt[s[i]-'a']++
	}
	result := ""
	for i := 0; i < len(charCnt); i++ {
		result += strconv.Itoa(charCnt[i]) + ","
	}
	return result
}

func main() {
	a := []string{
		"eat", "tea", "tan", "ate", "nat", "bat",
	}
	fmt.Println(groupAnagrams1(a))
}

File: leetcode/49-group-anagrams/main.py
"""
    1st attempt: hashtable + sort
    1 pass for iterating the words and put the words into corresponding hashtable
    1 pass for grouping the values from hashtable into the result
    Time O(n*klogk) n:number of words, k:length of charactors, klogk is due to the sorting
    Space O(nk)
    88 ms, faster than 99.89% 
"""


class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        ht = {}
        for s in strs:
            sortedStr = self.sortString(s)
            if sortedStr in ht:
                ht[sortedStr].append(s)
            else:
                ht[sortedStr] = [s]
        res = []
        for key in ht:
            res.append(ht[key])
        return res

    def sortString(self, s):
        a = sorted(s)
        return "".join(a)


print(Solution().groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

print("-----")

"""
    2nd attempt: hashtable but base on charactor count
    although it doesnt sort(nlogn), it runs slower than the 1st due to the string manipulation
    Time O(nk) n:number of words, k:length of charactors
    Space O(nk)

    120 ms, faster than 35.33%
"""


class Solution(object):
    def groupAnagrams(self, strs):
        ht = {}
        for s in strs:
            key = self.getSignature(s)
            if key not in ht:
                ht[key] = []
            ht[key].append(s)
        res = []
        for key in ht:
            res.append(ht[key])
        return res
            
        
    def getSignature(self, s):
        freqs = 26 * [0]
        for c in s:
            i = ord(c) - ord('a')
            freqs[i] += 1
        return tuple(freqs)


print(Solution().groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

print("-----")

"""
    3rd attempt: shorten the 2nd approach by just using tuples as keys

    Time O(nk) n:number of words, k:length of charactors
    Space O(nk)
    100 ms, faster than 54.16%
"""


class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        ht = {}
        for s in strs:
            structure = 26*[0]
            for c in s:
                i = ord(c)-ord('a')
                structure[i] += 1
            key = tuple(structure)
            if key not in ht:
                ht[key] = [s]
            else:
                ht[key].append(s)
        res = []
        for key in ht:
            res.append(ht[key])
        return res
